### PR TITLE
Add usalt fastq file linking, #336

### PR DESCRIPTION
**Pull request by patrikgrenfeldt**
This PR lets cg link microbial projects in housekeeper

### How to test linking of microbial sample:
0. install cg on stage on hasta, install servers on stage on hasta
1. us
1. manually add the sample fastqs to housekeeper `housekeeper add bundle SAMPLE_ID`, `housekeeper add file SAMPLE_FASTQ_PATH -t fastq`
1. link sample in cg `cg link-microbial SAMPLE_ID`

**Expected outcome:**
The fastq files and the links should have been listed

### How to test linking of an microbial order:
1. us
1. manually add the order sample fastqs to housekeeper `housekeeper add bundle SAMPLE_ID`, `housekeeper add file SAMPLE_FASTQ_PATH -t fastq`
1. link sample in cg `cg link-microbial MICROBIAL_ORDER_LIMS_ID`

**Expected outcome:**
The samples, fastq files and the links should have been listed

### How to test linking of MIP sample:
1. us
1. find a sample that is present in hk and is of this analysis type
1. link sample in cg `cg link SAMPLE_ID`

**Expected outcome:**
The fastq files and the links should have been listed

### How to test linking of an MIP case:
1. us
1. link sample in cg `cg link CASE_ID`

**Expected outcome:**
The samples, fastq files and the links should have been listed

### How to test linking of Balsamic sample:
1. us
1. find a sample that is present in hk and is of this analysis type
1. link sample in cg `cg link SAMPLE_ID`

**Expected outcome:**
The fastq files and the links should have been listed

### How to test linking of an Balsamic case:
1. us
1. link sample in cg `cg link CASE_ID`

**Expected outcome:**
The samples, fastq files and the links should have been listed

- [x] Code reviewed by @ingkebil 
- [x] Tested by @barrystokman 
- [x] Deploy signoff by @ingkebil 

Minor version bump: extended but did not change the public API.

### CODE REVIEW

**Comment by barrystokman** [comment]
> This PR lets cg link microbial projects in housekeeper
> 
> How to test linking of sample:
> 0. install cg on stage on hasta, install servers on stage on hasta

:heavy_check_mark: 
> 
> 1. us

:heavy_check_mark: 
> 2. manually add the sample fastqs to housekeeper `housekeeper add bundle SAMPLE_ID`, `housekeeper add file SAMPLE_FASTQ_PATH -t fastq`

`housekeeper add bundle ACC5498A1`
image

`housekeeper add file -t fastq ACC5498A1 /home/proj/stage/demultiplexed-runs/190412_A00689_0021_AHJM7LDSXX/Unaligned-Y151I10I10Y151/Project_990207/Sample_ACC5498A1/HJM7LDSXX_990207_S149_L001_R1_001.fastq.gz`
image

:heavy_check_mark: 

> 3. link sample in cg `cg link-microbial SAMPLE_ID`

`cg analysis link-microbial ACC5498A1`
image

image

:heavy_check_mark: 

> 
> Expected outcome:
> The fastq files and the links should have been listed

:heavy_check_mark: 
> 
> This PR lets cg link microbial projects in housekeeper
> 
> How to test linking of an microbial order:
> 0. install cg on stage on hasta, install servers on stage on hasta

:heavy_check_mark: 
> 
> 1. us

:heavy_check_mark: 
> 2. manually add the order sample fastqs to housekeeper `housekeeper add bundle SAMPLE_ID`, `housekeeper add file SAMPLE_FASTQ_PATH -t fastq`

`housekeeper add bundle ACC5484A8`
image

:heavy_check_mark: 

`housekeeper add file -t fastq ACC5484A8 /home/proj/stage/demultiplexed-runs/190412_A00689_0021_AHJM7LDSXX/Unaligned-Y151I10I10Y151/Project_632761/Sample_ACC5484A8/HJM7LDSXX_632761_S146_L001_R1_001.fastq.gz`

image

:heavy_check_mark: 

> 3. link sample in cg `cg link-microbial MICROBIAL_ORDER_LIMS_ID`

`cg analysis link-microbial --order ACC5484`

image

:heavy_check_mark: 
> 
> Expected outcome:
> The samples, fastq files and the links should have been listed

:heavy_check_mark:

File: cg/apps/usalt/fastq.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    FastqFileNameCreator: Creates valid usalt filenames
+    FastqFileConcatenator: Handles file concatenation
+    FastqHandler: Handles fastq file linking
+"""
+import logging
+import os
+from pathlib import Path
+from typing import List
+
+logger = logging.getLogger(__name__)
```
Comment (ingkebil) [inline cg/apps/usalt/fastq.py]:
```suggestion
LOGGER = logging.getLogger(__name__)
```

Comment (ingkebil) [reply cg/apps/usalt/fastq.py]: The global could be passed as an argument

File: cg/apps/usalt/fastq.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class FastqFileNameCreator:
+    """Creates valid usalt filename from the parameters"""
+
+    @staticmethod
+    def create(lane: str, flowcell: str, sample: str, read: str,
+               undetermined: bool = False) -> str:
+        """Name a FASTQ file following usalt conventions. Naming must be
+        xxx_R_1.fastq.gz and xxx_R_2.fastq.gz"""
```
Comment (ingkebil) [inline cg/apps/usalt/fastq.py]:
```suggestion

```

File: cg/apps/usalt/fastq.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+logger = logging.getLogger(__name__)
+
+
+class FastqFileNameCreator:
+    """Creates valid usalt filename from the parameters"""
+
+    @staticmethod
+    def create(lane: str, flowcell: str, sample: str, read: str,
+               undetermined: bool = False) -> str:
+        """Name a FASTQ file following usalt conventions. Naming must be
```
Comment (ingkebil) [inline cg/apps/usalt/fastq.py]:
```suggestion
        """Name a FASTQ file following usalt conventions. Naming must be sample_flowcell_lane_read.fastq.gz"""
```

File: cg/apps/usalt/fastq.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+class FastqHandler:
+    """Handles fastq file linking"""
+
+    def __init__(self, config):
+        self.root_dir = config['usalt']['root']
+
+    def link(self, case: str, sample: str, files: List):
+        """Link FASTQ files for a usalt sample.
+        Shall be linked to /<usalt root directory>/case-id/fastq/"""
```
Comment (ingkebil) [inline cg/apps/usalt/fastq.py]:
```suggestion
        Shall be linked to /<usalt root directory>/<case-id>/fastq/"""
```

File: cg/apps/usalt/fastq.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+        # The fastq files should be linked to /.../fastq/<project>/<sample>/*.fastq.gz.
+        wrk_dir = Path(self.root_dir) / 'fastq' / case / sample
+
+        wrk_dir.mkdir(parents=True, exist_ok=True)
+
+        linked_reads_paths = {1: [], 2: []}
```
Comment (ingkebil) [inline cg/apps/usalt/fastq.py]:
```suggestion

```
Unused

File: cg/apps/usalt/fastq.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            original_fastq_path = Path(fastq_data['path'])
+            linked_fastq_name = FastqFileNameCreator.create(
+                lane=fastq_data['lane'],
+                flowcell=fastq_data['flowcell'],
+                sample=sample,
+                read=fastq_data['read'],
+                undetermined=fastq_data['undetermined'],
+            )
+            linked_fastq_path = wrk_dir / linked_fastq_name
+            linked_reads_paths[fastq_data['read']].append(linked_fastq_path)
```
Comment (ingkebil) [inline cg/apps/usalt/fastq.py]:
```suggestion
```
unused

File: cg/apps/usalt/fastq.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    @staticmethod
+    def create(lane: str, flowcell: str, sample: str, read: str,
+               undetermined: bool = False) -> str:
+        """Name a FASTQ file following usalt conventions. Naming must be
+        xxx_R_1.fastq.gz and xxx_R_2.fastq.gz"""
+
+        # ACC1234A1_FCAB1ABC2_L1_1.fastq.gz sample_flowcell_lane_read.fastq.gz
+        flowcell = f"{flowcell}-undetermined" if undetermined else flowcell
+        return f"{sample}_{flowcell}_{lane}_{read}.fastq.gz"
```
Comment (ingkebil) [inline cg/apps/usalt/fastq.py]:
```suggestion
        return f"{sample}_{flowcell}_L{lane}_{read}.fastq.gz"
```

**Comment by ingkebil** [comment]
@patrikgrenfeldt this rewrite didn't break any tests! Only the linter is complaining :D

**Comment by patrikgrenfeldt** [comment]
@barrystokman, @ingkebil I think this one needs to be retested after the refactoring me and KB did last week, what do you say?

**Comment by ingkebil** [comment]
.. you should read your slack ;)

**Comment by barrystokman** [comment]
@patrikgrenfeldt  My little Artorias dude is already on the sticky :)
No changes to the old test cases? If not I'll get started right away.

**Comment by patrikgrenfeldt** [comment]
@barrystokman I'm sure we need to do `cg analysis link` for a MIP sample as well as for a Balsamic sample, right @ingkebil ?

**Comment by ingkebil** [comment]
Yes.

The test case should be expanded to include mip linking as well.

We also need unit tests for the mip linking.

**Comment by barrystokman** [comment]
Can't install on stage:

image

**Comment by barrystokman** [comment]
Retest TC1 Linking of microbial sample passed :heavy_check_mark: 

image

**Comment by barrystokman** [comment]
Retest TC2 Linking of microbial order passed :heavy_check_mark: 

image

**Comment by barrystokman** [comment]
> How to test linking of MIP sample:
> us
> find a sample that is present in hk and is of this analysis type

ACC5461A11

> link sample in cg cg link SAMPLE_ID

`cg link sample ACC5461A11`

> Expected outcome:
> The fastq files and the links should have been listed

image
@patrikgrenfeldt please double check test spec

**Comment by barrystokman** [comment]
> How to test linking of an MIP case:
> us
> link sample in cg cg link CASE_ID
@patrikgrenfeldt should be `cg analysis link -f CASE_ID`

family = modelpika

`cg analysis link -f crackmaggot`

> Expected outcome:
> The samples, fastq files and the links should have been listed

image

Test Passed :heavy_check_mark:

**Comment by barrystokman** [comment]
> How to test linking of Balsamic sample:
> us
> find a sample that is present in hk and is of this analysis type
> link sample in cg cg link SAMPLE_ID
> Expected outcome:
> The fastq files and the links should have been listed

Test data is copying... :sleeping:

**Comment by barrystokman** [comment]
> How to test linking of an Balsamic case:
> us
> link sample in cg cg link CASE_ID
> Expected outcome:
> The samples, fastq files and the links should have been listed

`cg analysis link -f fairsalmon`

image

Test Passed :heavy_check_mark:

**Pull request merged**